### PR TITLE
KEYCLOAK-12125 Introduce SameSite attribute in cookies

### PR DESCRIPTION
**Pull request by vmuzikar**

### CODE REVIEW

File: adapters/oidc/js/src/main/resources/login-status-iframe.html [hunk 1]
```diff
@@ -81,7 +81,16 @@
 
     function getCookie()
     {
-        var name = 'KEYCLOAK_SESSION=';
+        var cookie = getCookieByName('KEYCLOAK_SESSION');
+        if (cookie === null) {
+            cookie = getCookieByName('KEYCLOAK_SESSION_LEGACY');
```
Comment (jonkoops) [inline adapters/oidc/js/src/main/resources/login-status-iframe.html]: There is no need to re-assign `cookie` here, instead the result of `getCookieByName('KEYCLOAK_SESSION_LEGACY')` can be returned immediately.

**Comment by jonkoops** [comment]
JavaScript changes look good to me, can't say much about the rest.

File: adapters/oidc/js/src/main/resources/login-status-iframe.html (hunk 1 above)
Comment (vmuzikar) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: @jonkoops The legacy cookie is basically the same cookie as we set before this PR, i.e. it doesn't have the SameSite attribute. Chrome >80 won't send cookies in iframes that have no SameSite attribute set (which defaults to SameSite=Lax). In other words Chrome won't see the legacy cookie and at the same time some other browsers – especially older Apple browsers – reject the non-legacy cookie, so they see only the legacy cookie.

File: common/src/main/java/org/keycloak/common/util/ServerCookie.java [hunk 2]
```diff
@@ -32,6 +32,22 @@
     private static final String tspecials = ",; ";
     private static final String tspecials2 = "()<>@,;:\\\"/[]?={} \t";
 
+    public enum SameSiteAttributeValue {
+        NONE("None"), // we currently support only SameSite=None; this might change in the future
+        STRICT("Strict"),
```
Comment (abstractj) [inline common/src/main/java/org/keycloak/common/util/ServerCookie.java]: If we only support _NONE_ at the moment, I'd rather to make it a constant and not advertise LAX or STRICT in the code.

Comment (vmuzikar) [reply common/src/main/java/org/keycloak/common/util/ServerCookie.java]: @abstractj Done. (We actually had only the NONE value there at some point during the development. ;))

File: common/src/main/java/org/keycloak/common/util/ServerCookie.java [hunk 3]
```diff
@@ -173,7 +187,8 @@ public static void appendCookieValue(StringBuffer headerBuf,
                                          String comment,
                                          int maxAge,
                                          boolean isSecure,
-                                         boolean httpOnly) {
+                                         boolean httpOnly,
+                                         SameSiteAttributeValue sameSite) {
         StringBuffer buf = new StringBuffer();
```
Comment (abstractj) [inline common/src/main/java/org/keycloak/common/util/ServerCookie.java]: A minor nitpick, but StringBuilder is faster than StringBuffer, plus it seems that's our preferred choice into the codebase.

Comment (vmuzikar) [reply common/src/main/java/org/keycloak/common/util/ServerCookie.java]: @abstractj It was like this before, I wanted to change as little things as possible so I haven't touched this. Do you want me to switch it for StringBuffer?

File: adapters/oidc/js/src/main/resources/login-status-iframe.html [hunk 4]
```diff
@@ -81,7 +81,16 @@
...
+        }
+        return cookie;
+    }
+
+    function getCookieByName(name)
+    {
+        name = name + '=';
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: If you mean the separator `=`, we used that before so I didn't want to change that.
https://github.com/keycloak/keycloak/blob/05c428f6e74de7a544587537cebeee35aae70eb4/adapters/oidc/js/src/main/resources/login-status-iframe.html#L84

[the remark this answers is not in the document]
Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: Ah, yes I see now. My diff was messed up, thanks for clarifying.

File: adapters/oidc/js/src/main/resources/login-status-iframe.html (hunk 1 above)
Comment (vmuzikar) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: @jonkoops I'm sorry, I got it wrong the first time. You're correct, technically there's no need to re-assign the variable. But to me it seems more clear to have only one `return` but I don't really insist on it. Do you think this impacts performance?

File: common/src/main/java/org/keycloak/common/util/ServerCookie.java (hunk 3 above)
Comment (abstractj) [reply common/src/main/java/org/keycloak/common/util/ServerCookie.java]: @vmuzikar I overlooked this. That's not a deal breaker, we can leave it as is.

File: adapters/oidc/js/src/main/resources/login-status-iframe.html (hunk 1 above)
Comment (jonkoops) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: No apology needed, hahah. It's a code-style thing, so I would not worry about it too much.

In the JavaScript community early returns are generally preferred over re-assigning a variable. This is emphasised by the preferred usage of `const` over `var` and  `let` in modern codebases.

Comment (vmuzikar) [reply adapters/oidc/js/src/main/resources/login-status-iframe.html]: Ok, fair enough. ;) Updated per your suggestion.

**Comment by abstractj** [comment]
@vmuzikar @mhajas I ran some manual tests here and things are working as expected. I only have one question: is expected to get the warnings for SameSite on Chrome console even after applying these changes?

**Comment by vmuzikar** [comment]
@abstractj You mean the warning about setting a cookie without SameSite? I saw it somewhere before but I don't know why I cannot reproduce it right now. It is expected as we're still setting the legacy cookies that don't have SameSite.

**Review by pskopek — APPROVED** [review]
I have checked the latest changes and it looks fine. Approved.

**Comment by mhajas** [comment]
> @vmuzikar @mhajas I did some manual tests here and things are working as expected. I only have one question: is expected to get the warnings for SameSite on Chrome console even after applying these changes?

I saw that too. This might be caused by the fact we don't specify SameSite for each cookie, but only for these, we need it.

**Review by abstractj — APPROVED** [review]

**Review by abstractj — APPROVED** [review]
Even though the warnings are displayed on Chrome console, this is not a blocker to merge it as everything is working as expected.

**Comment by hmlnarik** [comment]
This PR also needs documentation

**Pull request merged**